lib/Dialect/Arc/Transforms/LegalizeStateUpdate.cpp

```diff
@@ -0,0 +1,428 @@
+    if (!accesses)
+      continue;
+
+    // Determine the states written by this op for which we have already seen a
+    // read earlier. These writes need to be legalized.
+    SmallVector<Value, 1> affectedStates;
+    for (auto access : accesses->accesses)
+      if (access.getInt() == AccessState::Write)
+        if (readStates.contains(access.getPointer()))
+          illegallyWrittenStates[access.getPointer()] = &op;
+
+    // Determine the states read by this op. This comes after handling of the
+    // writes, such that a block that contains both reads and writes to a state
+    // doesn't mark itself as illegal. Instead, we will descend into that block
+    // further down and do a more fine-grained legalization.
+    for (auto access : accesses->accesses)
+      if (access.getInt() == AccessState::Read)
+        readStates.insert(access.getPointer());
+  }
+
+  // Create a mapping from operations that create a read-after-write hazard to
+  // the states that they modify. Don't consider states that have already been
+  // legalized. This is important since we may have already created a temporary
+  // in a parent block which we can just reuse.
+  DenseMap<Operation *, SmallVector<Value, 1>> illegalWrites;
+  for (auto [state, op] : illegallyWrittenStates)
+    if (!legalizedStates.count(state))
+      illegalWrites[op].push_back(state);
+
+  // In a second forward pass over the block, insert the necessary temporary
+  // state to legalize the writes and recur into subblocks while providing the
+  // necessary rewrites.
+  SmallVector<Value> locallyLegalizedStates;
+
+  auto handleIllegalWrites = [&](Operation *op, SmallVector<Value, 1> &states) {
+    LLVM_DEBUG(llvm::dbgs() << "Visiting illegal " << op->getName() << "\n");
+
+    // Sort the states we need to legalize by a determinstic order establish
+    // during the access analysis. Without this the exact order in which states
+    // were moved into a temporary would be non-deterministic.
+    llvm::sort(states, [&](Value a, Value b) {
+      return analysis.stateOrder[a] < analysis.stateOrder[b];
+    });
+
+    // Legalize each state individually.
+    for (auto state : states) {
+      LLVM_DEBUG(llvm::dbgs() << "- Legalizing " << state << "\n");
+
+      // HACK: This is ugly, but we need a storage reference to allocate a state
+      // into. Ideally we'd materialize this later on, but the current impl of
+      // the alloc op requires a storage immediately. So try to find one.
+      Value storage;
+      if (auto allocOp = state.getDefiningOp<AllocStateOp>()) {
+        storage = allocOp.getStorage();
+      } else {
+        Block *currentBlock = block;
+        while (currentBlock && !storage) {
+          for (auto arg : currentBlock->getArguments())
+            if (arg.getType().isa<StorageType>())
+              storage = arg;
+          currentBlock = currentBlock->getParentOp()->getBlock();
+        }
```

Comment:
Couldn't we just check for AllocStateOp and RootInput and output an error otherwise, or is there another case that occurs currently? Or emit some warning that it just picks a storage by chance? Just silently picking any storage sounds like a long debugging session sometime in the future to me. 😬